cxl/lib/libcxl.c

```diff
@@ -8756,3 +8756,618 @@ CXL_EXPORT int cxl_memdev_fbist_test_randomsequence(struct cxl_memdev *memdev,
+	fprintf(stdout, "=========================== Read configuration file ============================\n");
+	fprintf(stdout, "Output Payload:");
+	for(int i=0; i<cmd->send_cmd->out.size; i++){
+		if (i % 16 == 0)
+		{
+			fprintf(stdout, "\n%04x  %02x ", i+offset, conf_read_out[i]);
+		}
+		else
+		{
+			fprintf(stdout, "%02x ", conf_read_out[i]);
+		}
+	}
+	fprintf(stdout, "\n");
+
+out:
+	cxl_cmd_unref(cmd);
+	return rc;
+	return 0;
+}
+
+#define CXL_MEM_COMMAND_ID_HCT_GET_CONFIG CXL_MEM_COMMAND_ID_RAW
+#define CXL_MEM_COMMAND_ID_HCT_GET_CONFIG_OPCODE 50689
+#define CXL_MEM_COMMAND_ID_HCT_GET_CONFIG_PAYLOAD_IN_SIZE 1
+#define CXL_MEM_COMMAND_ID_HCT_GET_CONFIG_PAYLOAD_OUT_SIZE 132
+
+struct cxl_mbox_hct_get_config_in {
+	u8 hct_inst;
+}  __attribute__((packed));
+
+struct cxl_mbox_hct_get_config_out {
+	u8 post_trig_depth;
+	u8 ignore_valid;
+	u8 rsvd;
+	u8 rsvd3;
+	__le32 trig_config[128];
+}  __attribute__((packed));
+
+CXL_EXPORT int cxl_memdev_hct_get_config(struct cxl_memdev *memdev,
+	u8 hct_inst)
+{
+	struct cxl_cmd *cmd;
+	struct cxl_mem_query_commands *query;
+	struct cxl_command_info *cinfo;
+	struct cxl_mbox_hct_get_config_in *hct_get_config_in;
+	struct cxl_mbox_hct_get_config_out *hct_get_config_out;
+	int rc = 0;
+	int trig_config_size;
+
+	cmd = cxl_cmd_new_raw(memdev, CXL_MEM_COMMAND_ID_HCT_GET_CONFIG_OPCODE);
+	if (!cmd) {
+		fprintf(stderr, "%s: cxl_cmd_new_raw returned Null output\n",
+				cxl_memdev_get_devname(memdev));
+		return -ENOMEM;
+	}
+
+	query = cmd->query_cmd;
+	cinfo = &query->commands[cmd->query_idx];
+
+	/* update payload size */
+	cinfo->size_in = CXL_MEM_COMMAND_ID_HCT_GET_CONFIG_PAYLOAD_IN_SIZE;
+	if (cinfo->size_in > 0) {
+		 cmd->input_payload = calloc(1, cinfo->size_in);
+		if (!cmd->input_payload)
+			return -ENOMEM;
+		cmd->send_cmd->in.payload = (u64)cmd->input_payload;
+		cmd->send_cmd->in.size = cinfo->size_in;
+	}
+
+	hct_get_config_in = (void *) cmd->send_cmd->in.payload;
+
+	hct_get_config_in->hct_inst = hct_inst;
+	rc = cxl_cmd_submit(cmd);
+	if (rc < 0) {
+		fprintf(stderr, "%s: cmd submission failed: %d (%s)\n",
+				cxl_memdev_get_devname(memdev), rc, strerror(-rc));
+		 goto out;
+	}
+
+	rc = cxl_cmd_get_mbox_status(cmd);
+	if (rc != 0) {
+		fprintf(stderr, "%s: firmware status: %d\n",
+				cxl_memdev_get_devname(memdev), rc);
+		rc = -ENXIO;
+		goto out;
+	}
+
+	if (cmd->send_cmd->id != CXL_MEM_COMMAND_ID_HCT_GET_CONFIG) {
+		 fprintf(stderr, "%s: invalid command id 0x%x (expecting 0x%x)\n",
+				cxl_memdev_get_devname(memdev), cmd->send_cmd->id, CXL_MEM_COMMAND_ID_HCT_GET_CONFIG);
+		return -EINVAL;
+	}
+
+	hct_get_config_out = (void *)cmd->send_cmd->out.payload;
+	fprintf(stdout, "==================== get hif/cxl trace buffer configuration ====================\n");
+	fprintf(stdout, "Post Trigger Depth: %x\n", hct_get_config_out->post_trig_depth);
+	fprintf(stdout, "Ignore Valid: %x\n", hct_get_config_out->ignore_valid);
+	// OPL size
+	trig_config_size = (cmd->send_cmd->out.size - 4) / 4;
```

Comment:
Add macro or variable for -4 / 4 that makes numbers meaningful e.g. (cmd->send_cmd->out.size - GET_CONFIG_OUT_NONVARIED_SIZE) / CONFIG_ENTRY_SIZE